File: ui-ngx/src/app/core/http/group-permission.service.ts
///
/// Copyright © 2016-2025 The Thingsboard Authors
///
/// Licensed under the Apache License, Version 2.0 (the "License");
/// you may not use this file except in compliance with the License.
/// You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software
/// distributed under the License is distributed on an "AS IS" BASIS,
/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
/// See the License for the specific language governing permissions and
/// limitations under the License.
///

import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { defaultHttpOptionsFromConfig, RequestConfig } from './http-utils';
import { PageLink } from '@shared/models/page/page-link';
import { PageData } from '@shared/models/page/page-data';
import { GroupPermission } from '@shared/models/group-permission.model';

@Injectable({
  providedIn: 'root'
})
export class GroupPermissionService {

  constructor(private http: HttpClient) {}

  public getGroupPermission(groupPermissionId: string, config?: RequestConfig): Observable<GroupPermission> {
    return this.http.get<GroupPermission>(`/api/groupPermission/${groupPermissionId}`, defaultHttpOptionsFromConfig(config));
  }

  public saveGroupPermission(groupPermission: GroupPermission, config?: RequestConfig): Observable<GroupPermission> {
    return this.http.post<GroupPermission>('/api/groupPermission', groupPermission, defaultHttpOptionsFromConfig(config));
  }

  public deleteGroupPermission(groupPermissionId: string, config?: RequestConfig): Observable<void> {
    return this.http.delete<void>(`/api/groupPermission/${groupPermissionId}`, defaultHttpOptionsFromConfig(config));
  }

  public getByUserGroup(userGroupId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<GroupPermission>> {
    return this.http.get<PageData<GroupPermission>>(
      `/api/userGroup/${userGroupId}/groupPermissions${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config)
    );
  }

  public getByEntityGroup(entityGroupId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<GroupPermission>> {
    return this.http.get<PageData<GroupPermission>>(
      `/api/entityGroup/${entityGroupId}/groupPermissions${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config)
    );
  }

  public getByRole(roleId: string, pageLink: PageLink, config?: RequestConfig): Observable<PageData<GroupPermission>> {
    return this.http.get<PageData<GroupPermission>>(
      `/api/role/${roleId}/groupPermissions${pageLink.toQuery()}`,
      defaultHttpOptionsFromConfig(config)
    );
  }
}

File: common/dao-api/src/main/java/org/thingsboard/server/dao/grouppermission/GroupPermissionService.java
/**
 * Copyright © 2016-2025 The Thingsboard Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingsboard.server.dao.grouppermission;

import org.thingsboard.server.common.data.id.EntityGroupId;
import org.thingsboard.server.common.data.id.GroupPermissionId;
import org.thingsboard.server.common.data.id.RoleId;
import org.thingsboard.server.common.data.id.TenantId;
import org.thingsboard.server.common.data.page.PageData;
import org.thingsboard.server.common.data.page.PageLink;
import org.thingsboard.server.common.data.permission.GroupPermission;
import org.thingsboard.server.dao.entity.EntityDaoService;

public interface GroupPermissionService extends EntityDaoService {

    GroupPermission findGroupPermissionById(TenantId tenantId, GroupPermissionId groupPermissionId);

    GroupPermission saveGroupPermission(GroupPermission groupPermission);

    void deleteGroupPermission(TenantId tenantId, GroupPermissionId groupPermissionId);

    PageData<GroupPermission> findGroupPermissionsByTenantId(TenantId tenantId, PageLink pageLink);

    PageData<GroupPermission> findGroupPermissionsByTenantIdAndUserGroupId(TenantId tenantId, EntityGroupId userGroupId, PageLink pageLink);

    PageData<GroupPermission> findGroupPermissionsByTenantIdAndEntityGroupId(TenantId tenantId, EntityGroupId entityGroupId, PageLink pageLink);

    PageData<GroupPermission> findGroupPermissionsByTenantIdAndRoleId(TenantId tenantId, RoleId roleId, PageLink pageLink);

    void deleteGroupPermissionsByTenantId(TenantId tenantId);

    void deleteGroupPermissionsByTenantIdAndUserGroupId(TenantId tenantId, EntityGroupId userGroupId);

    void deleteGroupPermissionsByTenantIdAndEntityGroupId(TenantId tenantId, EntityGroupId entityGroupId);

    void deleteGroupPermissionsByTenantIdAndRoleId(TenantId tenantId, RoleId roleId);

}
